File: practices/practice4.1/src/Pizza.cpp
#include "Pizza.hpp"
#include <print>
#include <algorithm>

Pizza::Pizza(const std::map<std::string, std::pair<double, bool>>& ingredients) {
    for (const auto& [ingredient, price_selected] : ingredients) {
        if (price_selected.second) {
            this->ingredients[ingredient] = {price_selected.first, price_selected.second, 1};
        }
    }
}

void Pizza::addIngredient(const std::string& ingredient, double price, bool required, int count) {
    if (ingredients.find(ingredient) != ingredients.end()) {
        std::get<COUNT_IDX>(ingredients[ingredient]) += count;
    } else {
        ingredients[ingredient] = {price, required, count};
    }
}

void Pizza::removeIngredient(const std::string &ingredient, int count) {
    std::get<COUNT_IDX>(ingredients[ingredient]) -= count;
    if (std::get<COUNT_IDX>(ingredients[ingredient]) <= 0) {
        ingredients.erase(ingredient);
    }
}

int Pizza::getIngredientCount(const std::string &ingredient) const {
    if (ingredients.find(ingredient) != ingredients.end()) {
        return std::get<COUNT_IDX>(ingredients.at(ingredient));
    }
    return 0;
}

void Pizza::showPizza() const {
    std::println("Pizza ingredients:");
    for (const auto& [ingredient, tuple] : ingredients) {
        std::println("Ingredient: {}, price: ${}, count - {}", ingredient, std::get<PRICE_IDX>(tuple), std::get<COUNT_IDX>(tuple));
    }
    std::println("Total price: ${}", calculatePrice());
}

double Pizza::calculatePrice() const {
    double price = 0;
    for (const auto& [ingredient, tuple] : ingredients) {
        price += std::get<PRICE_IDX>(tuple) * std::get<COUNT_IDX>(tuple);
    }
    return price;
}

File: homeworks/homework4.1/src/main/java/com/pizza/Pizza.java
package com.pizza;

import java.util.Map;
import java.util.HashMap;

public class Pizza {
    private final Map<String, Ingredient> ingredients = new HashMap<>();

    public Pizza(Map<String, Ingredient> ingredients) {
        ingredients.forEach((name, ingredient) -> {
            if (ingredient.isRequired()) {
                this.ingredients.put(name, ingredient);
            }
        });
    }

    public void addIngredient(String name, double price, boolean required, int count) {
        ingredients.putIfAbsent(name, new Ingredient(price, required, 0));
        Ingredient ingredient = ingredients.get(name);
        ingredient.addCount(count);
    }

    public void removeIngredient(String name, int count) {
        if (ingredients.containsKey(name)) {
            Ingredient ingredient = ingredients.get(name);
            ingredient.removeCount(count);
            if (ingredient.getCount() <= 0) {
                ingredients.remove(name);
            }
        }
    }

    public int getIngredientCount(String name) {
        return ingredients.getOrDefault(name, new Ingredient(0, false, 0)).getCount();
    }

    public void showPizza() {
        System.out.println("Pizza ingredients:");
        ingredients.forEach((name, ingredient) -> System.out.printf("Ingredient: %s, price: $%.2f, count - %d\n",
                name, ingredient.getPrice(), ingredient.getCount()));
        System.out.printf("Total price: $%.2f\n", calculatePrice());
    }

    public double calculatePrice() {
        return ingredients.values().stream()
                .mapToDouble(ingredient -> ingredient.getPrice() * ingredient.getCount())
                .sum();
    }
}
